Install components using flux helm-controller and source-controller

surajssd opened the pull request:
Fixes: https://github.com/kinvolk/lokomotive/issues/1528

surajssd committed:
Update vendor

Signed-off-by: Suraj Deshmukh <surajd.service@gmail.com>

Commit:
vendor: Add flux API dependency

- Add flux helm-controller API.
- Add flux source-controller API.
- Add sigs.k8s.io/controller-runtime.

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
Version: Capture additional git metadata

This commit collects more git metadata like branch and commit. This will
be further used to create GitRepository CR.

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
Add experimental subcommand

This commit adds support for `lokoctl experimental component apply`.

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
contour: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
contour: Make cert-generation job as hook

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
cert-manager: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
cert-manager: Mark CRs post-install

The CRDs and corresponding CRs are installed from the same template dir,
hence the initial validation fails since helm-controller cannot find the
schema on api-server. This commit makes the ordering little different by
installing the CRDs before CRs using helm hooks.

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
external-dns: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
httpbin: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
openebs-operator: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
openebs-storage-class: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
prometheus-operator: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
Add flux as control plane component

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
metrics-server: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
node-problem-detector: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
FLUO: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
inspektor-gadget: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
rook: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
rook-ceph: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
velero: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
Use library interacting with flux CRs

Commit:
metallb: Enable installation using helm-controller

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
components: Add Flux variables

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

knrt10 commented:
After talking with Suraj offline, it was thought that it would be better to move "converting components to Helm charts" commits in a different PR so that they can be merged to master.

knrt10 commented on pkg/components/aws-ebs-csi-driver/component.go:
```diff
@@ -17,6 +17,7 @@ package awsebscsidriver
 import (
 	"fmt"
 
+	api "github.com/fluxcd/helm-controller/api/v2beta1"
```
Just a doubt. Should we name it `api`? Won't it be little bit vague? How about something like `helmcontrollerapi`

surajssd replied:
changed this as well.

knrt10 commented on pkg/components/util/install.go:
```diff
@@ -19,25 +19,24 @@ import (
 	"github.com/kinvolk/lokomotive/pkg/version"
 )
 
+var fluxInstallInterval = metav1.Duration{Duration: 5 * time.Minute} //nolint:gomnd
```
This can be removed and instead we can use exported variable `components.FluxInstallInterval`

knrt10 commented on pkg/components/util/install.go:
```diff
@@ -47,7 +46,7 @@ func generateGitRepository() *sourceapi.GitRepository {
 		},
 		Spec: sourceapi.GitRepositorySpec{
-			Interval: metav1.Duration{5 * time.Minute},
+			Interval: fluxInstallInterval,
```
```suggestion
			Interval: components.FluxInstallInterval,
```

knrt10 reviewed (changes requested):
Nice work @surajssd. Some initial review

surajssd commented on assets/charts/control-plane/flux/templates/helm-controller.yaml:
```diff
@@ -0,0 +1,73 @@
+  labels:
+    app.kubernetes.io/instance: flux-system
+    app.kubernetes.io/part-of: flux
+    app.kubernetes.io/version: v0.16.2
```
This is upstream as in these configs were generated using the following command:

```bash
kubectl create ns flux-system --dry-run -o yaml
kubectl -n flux-system create rolebinding privileged-psp --clusterrole=privileged-psp --group=system:serviceaccounts:flux-system --dry-run -o yaml
flux install --version=latest --namespace=flux-system --components=source-controller,helm-controller --export
```

The generated configs were dumped into config files and voila created a chart out of it.

knrt10 replied:
Cool, then we can leave it as it is

Commit:
Release v0.9.0

Release CHANGELOG and scripts to update controller and worker nodes:
etcd and host kubelet.

Signed-off-by: Imran Pochi <imran@kinvolk.io>
Co-authored-by: Imran Pochi <imran@kinvolk.io>

Commit:
k8s: update to v1.21.4

Signed-off-by: Imran Pochi <imran@kinvolk.io>

Commit:
aks: update to v.1.21.2

Signed-off-by: Imran Pochi <imran@kinvolk.io>

Commit:
kubelet: remove Dockerfile, iscsiadm and update README

This commit removes the Dockerfile and iscsiadm files from the repo and
updates the README for newer instructions on how to build the kubelet
image.

Signed-off-by: Imran Pochi <imran@kinvolk.io>

Commit:
Revert "Merge pull request #1504 from kinvolk/surajssd/update-aws-csi-driver"

This reverts commit 004294c2f8e0395f293199be0108f63a346919f7, reversing
changes made to 9f67cb309b2f84ce8c06f2f289b4afd63f065759.

This is reverted due to flakiness/issue of `aws-ebs-csi-driver` on
Kubernetes version 1.21.4.

`ebs-plugin` container panics due to not being able to query instance
metadata:
```
W0909 10:57:15.342750       1 metadata.go:101] EC2 instance metadata is not available
panic: error getting information from metadata service or node object: node providerID empty, cannot parse

goroutine 1 [running]:
github.com/kubernetes-sigs/aws-ebs-csi-driver/pkg/driver.newNodeService(0xc00003a730, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/pkg/driver/node.go:88 +0x24c
github.com/kubernetes-sigs/aws-ebs-csi-driver/pkg/driver.NewDriver(0xc00050ff40, 0x7, 0x7, 0xc000078660, 0xc00032ec30, 0x1fded28)
	/go/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/pkg/driver/driver.go:94 +0x425
main.main()
	/go/src/github.com/kubernetes-sigs/aws-ebs-csi-driver/cmd/main.go:46 +0x212
```

Commit:
Add flag to conditionally create GitRepository

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>

Commit:
Merge pull request #1569 from kinvolk/imran/downgrade-aws-ebs-csi-driver

Revert "Merge pull request #1504 from kinvolk/surajssd/update-aws-csi-driver

Commit:
Merge pull request #1567 from kinvolk/imran/update-k8s-to-1.21.3

Update k8s to 1.21.4

Commit:
version: bump to v0.9.0+git

Signed-off-by: Imran Pochi <imran@kinvolk.io>

Commit:
version: bump to v0.9.0

Signed-off-by: Imran Pochi <imran@kinvolk.io>

Commit:
tinkerbell: Add btrfs-progs and gawk to Dockerfile

This commit adds additional packages which are required in the
`flatcar-install` script.

Signed-off-by: Imran Pochi <imran@kinvolk.io>

Commit:
Merge pull request #1570 from kinvolk/imran/update-tinkerbell-sandbox-image

tinkerbell: Add btrfs-progs and gawk to Dockerfile

Commit:
Merge pull request #1563 from kinvolk/release-v0.9.0

Release v0.9.0

surajssd commented on pkg/components/contour/component.go:
```diff
@@ -120,11 +132,54 @@ func (c *component) Metadata() components.Metadata {
+		Spec: api.HelmReleaseSpec{
+			Chart: api.HelmChartTemplate{
+				Spec: api.HelmChartTemplateSpec{
+					Chart: "./assets/charts/components/contour/",
```
Yep makes sense.

Commit:
Modify Component interface

This commit adds a new function in the Component interface that can
generate the "HelmRelease" file.

Signed-off-by: Suraj Deshmukh <suraj@kinvolk.io>
